core/dataset_next.cpp

```diff
@@ -0,0 +1,602 @@
+  if ((values.dimensions() != variances.dimensions()) ||
+      values.sparseDim() != variances.sparseDim())
+    throw std::runtime_error(
+        "Values and variances must have identical dimensions.");
+}
+
+/// Set (insert or replace) the data values with given name.
+///
+/// Throws if the provided values bring the dataset into an inconsistent state
+/// (mismatching dtype, unit, or dimensions).
+void Dataset::setValues(const std::string &name, Variable values) {
+  setDims(values.dims());
+  const auto it = m_data.find(name);
+  if (it != m_data.end() && it->second.variances) {
+    const auto &variances = *it->second.variances;
+    check_dtype(values, variances);
+    check_unit(values, variances);
+    check_dimensions(values, variances);
+  }
+  m_data[name].values = std::move(values);
+}
+
+/// Set (insert or replace) the data variances with given name.
+///
+/// Throws if the provided variances bring the dataset into an inconsistent
+/// state (mismatching dtype, unit, or dimensions).
+void Dataset::setVariances(const std::string &name, Variable variances) {
+  setDims(variances.dims());
+  const auto it = m_data.find(name);
+  if (it == m_data.end() || !it->second.values)
+    throw std::runtime_error("Cannot set variances: No data values for " +
+                             name + " found in dataset.");
+  const auto &values = *it->second.values;
+  check_dtype(values, variances);
+  check_unit(values, variances);
+  check_dimensions(values, variances);
+  m_data.at(name).variances = std::move(variances);
+}
+
+/// Set (insert or replace) the sparse coordinate with given name.
+///
+/// Sparse coordinates can exist even without corresponding data.
+void Dataset::setSparseCoord(const std::string &name, Variable coord) {
+  setDims(coord.dims());
+  if (!coord.isSparse())
+    throw std::runtime_error("Variable passed to Dataset::setSparseCoord does "
+                             "not contain sparse data.");
+  if (m_data.count(name)) {
+    const auto &data = m_data.at(name);
+    if ((data.values && (data.values->sparseDim() != coord.sparseDim())) ||
+        (!data.labels.empty() &&
+         (data.labels.begin()->second.sparseDim() != coord.sparseDim())))
+      throw std::runtime_error("Cannot set sparse coordinate if values or "
+                               "variances are not sparse.");
+  }
+  m_data[name].coord = std::move(coord);
+}
+
+/// Set (insert or replace) the sparse labels with given name and label name.
+void Dataset::setSparseLabels(const std::string &name,
+                              const std::string &labelName, Variable labels) {
+  setDims(labels.dims());
+  if (!labels.isSparse())
+    throw std::runtime_error("Variable passed to Dataset::setSparseLabels does "
+                             "not contain sparse data.");
+  if (m_data.count(name)) {
+    const auto &data = m_data.at(name);
+    if ((data.values && (data.values->sparseDim() != labels.sparseDim())) ||
```

Comment:
Looks to me that some of these checks would be better encapsulated and less error prone if in `DatasetData` itself.

Reply:
I just started changing this and notice that it is connected to the wider issue of giving exception safety guarantees. I suggest to tackle that together, I opened #248 for this.